### PR TITLE
Use push state to store exploding modes for conversations

### PR DESCRIPTION
**Pull request by buoyad**
cc @mmaxim r? @keybase/react-hackers

### CODE REVIEW

**Comment by mmaxim** [comment]
Suggestion in general for using this Gregor state in the UI: Why not store the actual Gregor state in Redux? I notice on `pushState`, the code just runs some actions and doesn't hold onto the state. Instead, you could have a specific place in the store for Gregor state, and then some functions which pull useful data out of it. This way you just have one place in the store for this kind of information, and don't need to essentially copy it into some other thing like in this PR. 

Thoughts?

**Comment by chrisnojima** [comment]
We should avoid having raw backend data types stored in our store. We want there to be a level of indirection so if that structure changes we don't have to change our side. The store is there to represent our data that we need to show the gui. Ideally we handle any specific data structures in the saga and it never escapes from there

**Comment by mmaxim** [comment]
In fact the Gregor state itself maps pretty cleanly onto the whole components/props world of React. The state is just a big JSON blob, the elements of which can map pretty cleanly onto the props of whatever components are controlling this exploding display.

**Comment by mmaxim** [comment]
In this case, the JS is creating the format, there is no "backend type" you need to worry about. You could modify my suggestion and filter the state for things you know about and only store that in the structure I described above.

**Comment by chrisnojima** [comment]
we also want to be able to disconnect what we write vs what we read. lets chat in person

**Comment by mmaxim** [comment]
Basically instead of something like this:
```
 const sawChatBanner = items.find(i => i.item && i.item.category === 'sawChatBanner')
 if (sawChatBanner) {
    yield Saga.put(TeamsGen.createSetTeamSawChatBanner())
  }
```
You'd have:
```
yield Saga.put(GregorGen.updateGregorState, filterForUI(items))
```

And in `mapStateToProps`:
```
showChatBanner: shouldShowChatBanner(state.gregor)
```

Just less boilerplate involved. Happy to chat in person.

**Comment by buoyad** [comment]
I think in this particular example we are pretty much doing what you describe. We're not letting go of any of the data, just mapping it onto a part of the store (`state.chat2.explodingModes`) that is conveniently poised to control the different parts of the UI & actions that need it to make decisions.

**Comment by mmaxim** [comment]
Yeah I'm not saying you are just dropping the data altogether, it just gets dispersed all over the place. I am just arguing you store the Gregor state by itself, and then pull the stuff out of it when updating props whatever components care about it. In other words, you don't need a custom action for every item in the Gregor state you care about.

**Comment by mmaxim** [comment]
Anyway, just a suggestion, don't want to hang anything up on it, but I think it might make interacting with Gregor a little more enjoyable.

File: shared/actions/chat2/index.js [hunk 1]
```diff
@@ -1893,6 +1894,58 @@ const setConvRetentionPolicy = (action: Chat2Gen.SetConvRetentionPolicyPayload)
...
+const setConvExplodingMode = (action: Chat2Gen.SetConvExplodingModePayload) => {
+  const {conversationIDKey, seconds} = action.payload
+  logger.info(`Setting exploding mode for conversation ${conversationIDKey} to ${seconds}`)
+  const cat = Constants.explodingModeGregorKey(conversationIDKey)
+  if (seconds === 0) {
+    // dismiss the category so we don't leave cruft in the push state
+    return Saga.call(RPCTypes.gregorDismissCategoryRpcPromise, {category: cat})
+  }
+  return Saga.call(RPCTypes.gregorInjectItemRpcPromise, {
+    body: seconds.toString(),
```
Comment (mmaxim) [inline shared/actions/chat2/index.js]: Why toString this?

Comment (buoyad) [reply shared/actions/chat2/index.js]: `seconds` is a `number`, and `gregorInjectItem` takes a body that is a `string`

Comment (mmaxim) [reply shared/actions/chat2/index.js]: Ah I guess it forces you to give a string.

**Comment by buoyad** [comment]
@keybase/react-hackers bump!

File: shared/actions/chat2/index.js [hunk 2]
```diff
@@ -1893,6 +1894,58 @@ const setConvRetentionPolicy = (action: Chat2Gen.SetConvRetentionPolicyPayload)
   }
   return ret
 }
+
+const setConvExplodingMode = (action: Chat2Gen.SetConvExplodingModePayload) => {
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: lets talk to @mmaxim about how this works offline etc

**Comment by buoyad** [comment]
Waiting on CI fix to merge

**Pull request merged**